File: util_cpp/list.cpp
#include "list.h"
using namespace std;

bool is_equal_list(ListNode *l1, ListNode *l2)
{
    while (true) {
        if (!l1 && !l2)
            return true;
        if (!l1 || !l2)
            return false;
        if (l1->val != l2->val)
            return false;
        l1 = l1 ? l1->next : l1;
        l2 = l2 ? l2->next : l2;
    }
}

void print_list(ListNode *list)
{
    while(list) {
        std::cout << list->val << " ";
        list = list->next;
    }
    std::cout << std::endl;
}

ListNode* build_from_vector(vector<int> vec)
{
    ListNode* dummy = new ListNode(0);
    ListNode* p = dummy;
    for (int n : vec) {
        p->next = new ListNode(n);
        p = p->next;
    }
    return dummy->next;
}

vector<int> build_vector(ListNode* list)
{
    vector<int> res;
    ListNode *p = list;
    while (p) {
        res.push_back(p->val);
        p = p->next;
    }
    return res;
}

File: utils_js/list.js

function ListNode(val, next) {
  this.val = (val===undefined ? 0 : val);
  this.next = (next===undefined ? null : next);
}

/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {Boolean}
 */
function isEqualList(l1, l2) {
  let p1 = l1;
  let p2 = l2;
  while (p1 && p2) {
    if (p1.val != p2.val)
      return false;
    p1 = p1.next;
    p2 = p2.next;
  }
  if (p1 || p2)
    return false;
  return true;
}

/**
 * @param {Array} array
 * @return {ListNode}
 */
function buildList(array) {
  let dummyHead = new ListNode(0);
  let p = dummyHead;
  for (let i = 0; i < array.length; i++) {
    p.next = new ListNode(array[i]);
    p = p.next;
  }
  return dummyHead.next;
}

function printList(list) {
  let p = list;
  while (p) {
    console.log(p.val);
    p = p.next;
  }
}

const ListUtils = {
  ListNode,
  isEqualList,
  printList,
  buildList,
};

export default ListUtils;